src/hotspot/share/opto/library_call.cpp

```diff
@@ -5206,24 +5206,102 @@ bool LibraryCallKit::inline_vectorizedMismatch() {
-  set_result(result);
+    if (enable_pi &&
+        Matcher::match_rule_supported_vector(Op_VectorMaskGen , vec_len, elem_bt) &&
+        Matcher::match_rule_supported_vector(Op_LoadVectorMasked , vec_len, elem_bt) &&
+        Matcher::match_rule_supported_vector(Op_VectorCmpMasked, vec_len, elem_bt)) {
+
+      Node* length_in_bytes = _gvn.transform(new LShiftINode(length, scale));
+      Node* pi_size = intcon(ArrayOperationPartailInlineSize);
+      Node* length_cmp = _gvn.transform(new CmpINode(length_in_bytes, pi_size));
+      Node* cmp_res = _gvn.transform(new BoolNode(length_cmp, BoolTest::le));
+
+      fast_path = generate_guard(cmp_res, NULL, PROB_MAX);
+
+      const TypeVect* vt = TypeVect::make(elem_bt, vec_len);
+      Node* mask_gen = _gvn.transform(new VectorMaskGenNode(ConvI2L(length), TypeVect::VECTMASK, elem_bt));
+
+      const TypePtr* ptr_type_a = obja_adr->Value(&_gvn)->isa_ptr();
+      const TypePtr* ptr_type_b = objb_adr->Value(&_gvn)->isa_ptr();
+
+      int alias_idx = C->get_alias_index(top_a);
+      Node* mm = memory(alias_idx);
+      Node* masked_load1 = _gvn.transform(new LoadVectorMaskedNode(fast_path, mm, obja_adr,
+                                                                   ptr_type_a, vt, mask_gen));
+      alias_idx = C->get_alias_index(top_b);
+      mm = memory(alias_idx);
+      Node* masked_load2 = _gvn.transform(new LoadVectorMaskedNode(fast_path, mm, objb_adr,
+                                                                   ptr_type_b, vt, mask_gen));
+
+      fastcomp_result = _gvn.transform(new VectorCmpMaskedNode(masked_load1, masked_load2,
+                                                               mask_gen, TypeInt::INT));
+      C->set_max_vector_size(MAX2((uint)ArrayOperationPartailInlineSize, C->max_vector_size()));
+    }
+  }
+
+  if (!stopped()) {
+    Node* call = make_runtime_call(RC_LEAF,
+                                   OptoRuntime::vectorizedMismatch_Type(),
+                                   stubAddr, stubName, TypePtr::BOTTOM,
+                                   obja_adr, objb_adr, length, scale);
+
+    Node* call_result = _gvn.transform(new ProjNode(call, TypeFunc::Parms));
+    Node* call_mem = map()->memory();
+
+    Node* exit_block = new RegionNode(3);
+    exit_block->init_req(1, fast_path);
+    exit_block->init_req(2, control());
+    exit_block = _gvn.transform(exit_block);
+
+    Node* result = new PhiNode(exit_block, TypeInt::INT);
+    result->init_req(1, fastcomp_result);
+    result->init_req(2, call_result);
+    result = _gvn.transform(result);
+
+    Node* mem_phi = new PhiNode(exit_block, Type::MEMORY, TypePtr::BOTTOM);
+    mem_phi->init_req(1, init_mem);
+    mem_phi->init_req(2, call_mem);
+
+    set_all_memory(_gvn.transform(mem_phi));
+    set_control(exit_block);
+    set_result(result);
+  } else {
+    set_control(fast_path);
+    set_result(fastcomp_result);
+  }
+  clear_upper_avx();
```

Comment:
There was no `clear_upper_avx()` before. 
Was it overlooked before or is it needed only for new code (when partial inlining takes place)?

Reply:
This is added for newer code dealing with vectors.